### PR TITLE
Bug 1972383: openshift authorization proxy: escape header key values

### PR DESCRIPTION
**Pull request by stlaz**
The authorization.openshift.io RBAC proxy is taking values from
the Extra field of a UserInfo of a user that made a request against
this API. The fields of the map may generally contain values that
cannot appear in an HTTP header, like '/' that's commonly separating
an API annotation from the API version or resource.

Copy the k8s.io header escaping function and use it in the proxy
that's handling the OpenShift authorization API.

/cc @s-urbaniak @sttts 

/cherry-pick release-4.7

### CODE REVIEW

File: pkg/client/impersonatingclient/impersonate.go [hunk 1]
```diff
@@ -86,3 +88,110 @@ func (c *impersonatingRESTClient) APIVersion() schema.GroupVersion {
...
+	return int(b) < len(legalHeaderKeyBytes) && legalHeaderKeyBytes[b]
+}
+
+func shouldEscape(b byte) bool {
+	// url.PathUnescape() returns an error if any '%' is not followed by two
+	// hexadecimal digits, so we'll intentionally encode it.
+	return !legalHeaderByte(b) || b == '%'
+}
+
+func headerKeyEscape(key string) string {
```
Comment (s-urbaniak) [inline pkg/client/impersonatingclient/impersonate.go:102]: can we add a reference to client-go?

Comment (stlaz) [reply pkg/client/impersonatingclient/impersonate.go:103]: added

**Comment by osherdp** [comment]
/lgtm
works on my env

**Pull request merged**